### PR TITLE
lapack/gonum: fix bugs in Dgesvd and rewrite DgesvdTest

### LINKED ISSUE #816: lapack/gonum: coverage of Dgesvd is insufficient
Based on the recently found issues with ldwork I tried to modify DgesvdTest to include also minimum and medium workspace lengths. I get the following panic which is in code not currently covered by the test:
```
panic: lapack: tau has insufficient length [recovered]
	panic: lapack: tau has insufficient length

goroutine 18 [running]:
testing.tRunner.func1(0xc0000c2100)
	/usr/lib/go/src/testing/testing.go:792 +0x387
panic(0x678ce0, 0x72f0b0)
	/usr/lib/go/src/runtime/panic.go:513 +0x1b9
gonum.org/v1/gonum/lapack/gonum.Implementation.Dormqr(0x4e52, 0x9, 0x5, 0x5, 0xc0000e5450, 0x19, 0x19, 0x5, 0xc0000e5618, 0x14, ...)
	/home/chalupecky/src/gonum.org/v1/gonum/lapack/gonum/dormqr.go:67 +0x16e5
gonum.org/v1/gonum/lapack/gonum.Implementation.Dormbr(0x4e5251, 0x9, 0x5, 0x5, 0xc0000e5450, 0x19, 0x19, 0x5, 0xc0000e5618, 0x14, ...)
	/home/chalupecky/src/gonum.org/v1/gonum/lapack/gonum/dormbr.go:118 +0x58d
gonum.org/v1/gonum/lapack/gonum.Implementation.Dgesvd(0xc000064141, 0x9, 0x5, 0xc000122000, 0x2d, 0x2d, 0x5, 0xc00007a7e0, 0x5, 0x5, ...)
	/home/chalupecky/src/gonum.org/v1/gonum/lapack/gonum/dgesvd.go:788 +0x3bd1
gonum.org/v1/gonum/lapack/testlapack.DgesvdTest(0xc0000c2100, 0x731860, 0x94a7a8)
```
It happens when lwork is the documented minimum.

### PR DESCRIPTION
**Pull request by vladimir-ch**
Closes #816 

Please take a look.

### CODE REVIEW

**Comment by vladimir-ch** [comment]
The coverage is now practically 100%, including the rescaling. I'll try to do something similar for the GSVD, I think I even promised doing something like this in the past.

File: blas/blas64/blas64.go [hunk 1]
```diff
@@ -29,8 +29,8 @@ func Implementation() blas.Float64 {
 // Vector represents a vector with an associated element increment.
 type Vector struct {
 	N    int
-	Inc  int
```
Comment (kortschak) [inline blas/blas64/blas64.go]: Wasn't this already changed? Did I confabulate that?

File: lapack/testlapack/dgesvd.go [hunk 2]
```diff
@@ -21,251 +23,351 @@ type Dgesvder interface {
...
+		aNorm = 1
+	case 3, 4, 5:
+		// Scaled random matrix.
+		// Generate singular values.
+		s := make([]float64, minmn)
+		// 4 means that s[i] = 1 - i*(1-1/cond)/(minmn-1),
+		// where cond = max(1, minmn),
+		// false means that signs of s[i] are not randomly flipped,
+		// 1 means that the random numbers are drawn uniformly from [0,1).
+		Dlatm1(s, 4, float64(max(1, minmn)), false, 1, rnd)
```
Comment (kortschak) [inline lapack/testlapack/dgesvd.go]: You could do this as (only suggestion):
```
Dlatm1(s,
    4,                      // s[i] = 1 - i*(1-1/cond)/(minmn-1),
    float64(max(1, minmn)), // where cond = max(1, minmn)
    false,                  // signs of s[i] are not randomly flipped
    1, rnd)                 // random numbers are drawn uniformly from [0,1)
```

File: blas/blas64/blas64.go (hunk 1 above)
Comment (kortschak) [reply blas/blas64/blas64.go]: I recall now, that was at call/use sites.

File: lapack/testlapack/dgesvd.go [hunk 3]
```diff
@@ -21,251 +23,351 @@ type Dgesvder interface {
...
+		for i := range work {
+			work[i] = math.NaN()
+		}
 
-		work := make([]float64, 1)
-		impl.Dgesvd(jobU, jobVT, m, n, a, lda, s, u, ldu, vt, ldvt, work, -1)
+		// Compute the full SVD which will be used later for checking the partial results.
+		ok := impl.Dgesvd(lapack.SVDAll, lapack.SVDAll, m, n, a, lda, sAll, uAll, ldu, vtAll, ldvt, work, len(work))
+		if !ok {
+			t.Fatalf("Case %v: unexpected failure in full SVD", prefix, a, sAll)
```
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply lapack/testlapack/dgesvd.go]: Thanks, this was a leftover from debugging before I identified the missing ok returns.

File: lapack/testlapack/dgesvd.go (hunk 2 above)
Comment (vladimir-ch) [reply lapack/testlapack/dgesvd.go]: Nice, thanks.

One day I will split `Dlatm1` into several smaller single-purpose functions or at least unexport it. The signature is really bad.

**Comment by vladimir-ch** [comment]
PTAL

**Comment by vladimir-ch** [comment]
>  I'll try to do something similar for the GSVD, I think I even promised doing something like this in the past.

#679 is the relevant ticket.

**Pull request merged**